plugins/bcm57xx/fu-bcm57xx-device.c

```diff
@@ -0,0 +1,767 @@
+	img_stage1 = fu_firmware_get_image_by_id (firmware_tmp, "stage1", NULL);
+	if (img_stage1 != NULL)
+		fu_firmware_add_image (firmware, img_stage1);
+	img_stage2 = fu_firmware_get_image_by_id (firmware_tmp, "stage2", NULL);
+	if (img_stage2 != NULL)
+		fu_firmware_add_image (firmware, img_stage2);
+	img_ape = fu_firmware_get_image_by_id (firmware_tmp, "ape", NULL);
+	if (img_ape != NULL)
+		fu_firmware_add_image (firmware, img_ape);
+
+	/* success */
+	return g_steal_pointer (&firmware);
+}
+
+static gboolean
+fu_bcm57xx_device_write_firmware (FuDevice *device,
+				  FuFirmware *firmware,
+				  FwupdInstallFlags flags,
+				  GError **error)
+{
+	FuBcm57xxDevice *self= FU_BCM57XX_DEVICE (device);
+	const guint8 *buf;
+	gsize bufsz = 0;
+	gsize bufsz_dwrds = BCM_FIRMWARE_SIZE / sizeof(guint32);
+	g_autofree guint32 *buf_dwrds = g_new0 (guint32, bufsz_dwrds);
+	g_autoptr(FuDeviceLocker) locker = NULL;
+	g_autoptr(GBytes) blob = NULL;
+
+	/* build the images into one linear blob of the correct size */
+	fu_device_set_status (device, FWUPD_STATUS_DECOMPRESSING);
+	blob = fu_firmware_write (firmware, error);
+	if (blob == NULL)
+		return FALSE;
+
+	/* align into uint32_t buffer */
+	buf = g_bytes_get_data (blob, &bufsz);
+	for (guint i = 0; i < bufsz_dwrds; i++) {
+		if (!fu_memcpy_safe ((guint8 *) buf_dwrds,
+				     bufsz_dwrds * sizeof(guint32),
+				     i * sizeof(guint32),		/* dst */
+				     buf, bufsz, i * sizeof(guint32),	/* src */
+				     sizeof(guint32), error))
+			return FALSE;
+	}
+
+	/* hit hardware */
+	fu_device_set_status (device, FWUPD_STATUS_DEVICE_WRITE);
+	locker = fu_device_locker_new_full (self,
+					    (FuDeviceLockerFunc) fu_bcm57xx_device_nvram_acquire_lock,
+					    (FuDeviceLockerFunc) fu_bcm57xx_device_nvram_release_lock,
+					    error);
+	if (locker == NULL)
+		return FALSE;
+	if (!fu_bcm57xx_device_nvram_set_mode (self, FU_BCM57XX_NVRAM_MODE_ENABLED_WRITE, error))
+		return FALSE;
+	if (!fu_bcm57xx_device_nvram_write (self, 0x0, buf_dwrds, bufsz_dwrds, error))
+		return FALSE;
+	if (!fu_bcm57xx_device_nvram_set_mode (self, FU_BCM57XX_NVRAM_MODE_DISABLED, error))
+		return FALSE;
+
+	/* reset APE */
+	if (!fu_bcm57xx_device_reset_ape (self, error))
+		return FALSE;
+
+	/* success */
+	return TRUE;
+}
+
+static gboolean
+fu_bcm57xx_device_ethtool_read (FuBcm57xxDevice *self,
+				guint32 address, guint32 *buf, guint32 bufsz_wrds,
+				GError **error)
+{
+#ifdef HAVE_ETHTOOL_H
+	gsize eepromsz;
+	gint rc = -1;
+	struct ethtool_drvinfo drvinfo = { 0 };
+	struct ifreq ifr = { 0 };
+	g_autofree struct ethtool_eeprom *eeprom = NULL;
+
+	/* get driver info */
+	drvinfo.cmd = ETHTOOL_GDRVINFO;
+	strncpy (ifr.ifr_name, self->ethtool_iface, IFNAMSIZ - 1);
+	ifr.ifr_data = (char *) &drvinfo;
+#ifdef HAVE_IOCTL_H
+	rc = ioctl (self->ethtool_fd, SIOCETHTOOL, &ifr);
+#endif
+	if (rc < 0) {
+		g_set_error (error,
+			     G_IO_ERROR,
+			     G_IO_ERROR_FAILED,
+			     "cannot get driver information [%i]", rc);
```

Comment:
if you have ethtool and don't have ioctl.h, this error code will be wrong.  maybe the `#ifdef` needs `#else` clause for that.

Reply:
Ya, I did think it would just call back to rc==-1, but I agree it's more explicit with a #else